include/deal.II/fe/fe_coupling_values.h

```diff
@@ -0,0 +1,653 @@
+     * associated with a DoFHandler object (as the function above does), this
+     * function instead takes these local nodal values through its first
+     * argument.
+     *
+     * @param[in] dof_values A vector of local nodal values. This vector must
+     *   have a length equal to the size of the dof renumbering vector.
+     *
+     * @param[out] gradients A vector of gradients of the given finite element
+     *   field, at the renumbered quadrature points on the current object.
+     *
+     * @tparam InputVector The @p InputVector type must allow creation of an
+     *   ArrayView object from it; this is satisfied by the `std::vector` class,
+     *   among others.
+     *
+     * @dealiiRequiresUpdateFlags{update_gradients}
+     */
+    template <class InputVector>
+    void
+    get_function_gradients_from_local_dof_values(
+      const InputVector &dof_values,
+      std::vector<solution_gradient_type<typename InputVector::value_type>>
+        &gradients) const;
+
+  private:
+    /**
+     * The data structure that stores the renumbering of the degrees of freedom
+     * and of the quadrature points.
+     */
+    const RenumberingData &data;
+
+    /**
+     * Produce an inner vector compatible with the inner view, after copying the
+     * values with the correct numbering from the outer vector.
+     */
+    template <typename InputVector>
+    const InputVector &
+    outer_to_inner_dofs(const InputVector &outer_vector) const;
+
+    /**
+     * Produce an inner vector compatible with the inner view, and zero out its
+     * entries if necessary.
+     */
+    template <typename ValueType>
+    std::vector<ValueType> &
+    outer_to_inner_values(std::vector<ValueType> &outer_values) const;
+
+    /**
+     * Return the outer argument renumbered according to the quadrature
+     * renumbering. The values in the inner values are copied to the right
+     * position in the outer vector.
+     */
+    template <typename ValueType>
+    void
+    inner_to_outer_values(const std::vector<ValueType> &inner_values,
+                          std::vector<ValueType>       &outer_values) const;
+
+    /**
+     * Store a reference to the underlying view.
+     */
+    const ViewType &view;
+  };
+} // namespace FEValuesViews
+
+
+#ifndef DOXYGEN
+
+
+/*------------------------ Inline functions: namespace FEValuesViews --------*/
+
+namespace FEValuesViews
+{
+  template <typename ViewType>
+  RenumberedView<ViewType>::RenumberedView(const ViewType        &view,
+                                           const RenumberingData &data)
+    : view(view)
+    , data(data)
+  {}
+
+
+
+  template <typename Number>
+  inline std::string
+  RenumberingData::get_unique_container_name(
+    const std::string &prefix,
+    const unsigned int size,
+    const Number      &exemplar_number) const
+  {
+    return prefix + "_" + Utilities::int_to_string(size) + "_" +
+           Utilities::type_to_string(exemplar_number);
+  }
```

Comment:
This function is part of the `public` interface of the class. Looking through where it is used, it seems like it's only actually used in `RenumberedView`. What do you think about making the function `private` and making `RenumberedView` a friend? This seems to be an implementation detail that does not need to be part of the public interface of the structure.